pkg/controller/certificates/trigger/policies/gatherer_test.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policies
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/utils/pointer"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmlisters "github.com/jetstack/cert-manager/pkg/client/listers/certmanager/v1"
+)
+
+func TestDataForCertificate(t *testing.T) {
+	tests := []struct {
+		name                       string
+		mockSecretLister           func(t *testing.T) secretListerMock
+		mockListRequests           func(t *testing.T) requestListerMock
+		givenCrt                   *cmapi.Certificate
+		wantCurrentRevisionRequest *cmapi.CertificateRequest
```

Comment:
Are we also checking that the returned Certificate is the same that we are expecting?
Moving to the Input struct will help here.

Reply:
We do check that the returned Certificate is the same; and since we do that check for every single test case (and that using Input would require me to copy the gen.CertificateFrom twice in the test case), I figured I would:
- instead of using `input` directly as a field, have the fields `wantSecret` and `wantCurrentRevisionRequest `;
- and check that `givenCrt` and the certificate given by DataForCertificate match in the `t.Run` itself

Reply:
When we also add the LastFailedRequest field re: https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1611141485045400

which will need testing, I think it makes sense to make this a test field by itself.

Reply:
> I think it makes sense to make Input a test field by itself.

Definitely +1, although it'd be really great if we can break this PR up so we aren't making sweeping refactors to tests at the same time as adding new functionality. Can we refactor these tests *before* change the actual implementation, so we can be sure that tests that previously passed continue to pass, and we have not introduced any regressions?

I'm particularly conscious here simply because a mix-up could cause all certificates to be endlessly re-issued for users 😅

Reply:
:+1: 

Sorry, wasn't advocating making those changes in this PR . Should have been more clear.

Reply:
No need to be sorry! I thought I'd just emphasise it because changes to this area of code can be scary 🙈